File: archivelib-sys2/c-lib/src/r_compress/fn199.cpp

#include "r_compress.hpp"

void RCompress::fn199(int16_t uncompressed_buffer_index200, int16_t _201) {
  uint8_t *_451;
  uint8_t *l_uncompressed_buffer278;
  int16_t _226, _452, _204, _453;
  _452 = MAX_COMPRESSION_CYCLES;
  data->dat168 = 0;
  _451 = &data->uncompressed_buffer[uncompressed_buffer_index200];
  _204 = _201;
  while ((_204 = data->dat_arr163[_204]) != TRUE157) {
    if (--_452 < 0)
      break;
    l_uncompressed_buffer278 = &data->uncompressed_buffer[_204];
    if (_451[data->dat168] != l_uncompressed_buffer278[data->dat168])
      continue;
    if (_451[0] != l_uncompressed_buffer278[0])
      continue;
    if (_451[1] != l_uncompressed_buffer278[1])
      continue;
    if (_451[2] != l_uncompressed_buffer278[2])
      continue;
    for (_226 = 3; _226 < CONST_N140_IS_256; _226++)
      if (_451[_226] != l_uncompressed_buffer278[_226])
        break;
    if (_226 > data->dat168) {
      _453 = (int16_t)(uncompressed_buffer_index200 - _204 - 1);
      if (_453 < 0)
        _453 += data->max_input_data_size;
      if (_453 >= data->max_input_data_size) {
        break;
      }
      data->dat169 = _453;
      if ((data->dat168 = _226) >= CONST_N140_IS_256)
        break;
    }
  }
}

File: archivelib-rs-tdd/src/compress/fn199.rs
use crate::compress::RCompressData;
use crate::consts::{MAX_COMPRESSION_CYCLES, MAX_RUN_LENGTH140};
use std::io::{Read, Write};

impl<R: Read, W: Write> RCompressData<R, W> {
  pub fn fn199(&mut self, uncompressed_buffer_index200: i16, var201: i16) {
    let mut test_idx = var201 as usize;
    let orig_idx = uncompressed_buffer_index200 as usize;
    self.dat168 = 0 as i16;
    for _ in 0..=MAX_COMPRESSION_CYCLES {
      if self.dat_arr163[test_idx] == -1 {
        break;
      }
      test_idx = self.dat_arr163[test_idx] as usize;
      if self.uncompressed_buffer[orig_idx + self.dat168 as usize]
        != self.uncompressed_buffer[test_idx + self.dat168 as usize]
      {
        continue;
      }
      let run_end = 0;
      for run_end in 0..MAX_RUN_LENGTH140 {
        if self.uncompressed_buffer[orig_idx + run_end]
          != self.uncompressed_buffer[test_idx + run_end]
        {
          break;
        }
      }
      if run_end < 3 {
        continue;
      }
      if run_end > self.dat168 {
        let mut offset = (orig_idx - test_idx - 1) as i16;
        if offset < 0 {
          offset = offset + self.max_uncompressed_data_size as i16
        }
        if offset >= self.max_uncompressed_data_size as i16 {
          break;
        }
        self.dat169 = offset;
        self.dat168 = run_end;
        if self.dat168 >= MAX_RUN_LENGTH140 as i16 {
          break;
        }
      }
    }
  }
}
